### PR TITLE
fix(zones-locations): fix zone locations updates

### PR DESCRIPTION
**Pull request by qmateub**
affects: @commercetools/sync-actions

#### Summary

This PR fixes some issues that appear during the UX/UAT that were indeed related to the sync-actions of zones.

I've added some code comment to explaining the use cases. Hope that makes sense 👍 

#### Description
Basically there were some cases that the sync-actions were not handling correctly guiding us to some UI errors (API errors) in the MC FE.

Working example. Let's say I have this object:

`locations: [{country: 'Spain'} ,{country: 'Germany'}]`

And my new object is:

`locations: [{country: 'Germany'}]`

The sync actions builded were:
```js
  [
    {action: "removeLocation", location: {country: "Spain"}},
    {action: "addLocation", location: {country: "Germany"}},
    {action: "removeLocation", location: {country: "Germany"}}
  ] 
```

That set of actions leads us to an API error because we can't add a location that is already assigned to the zone (Germany).

### CODE REVIEW

File: packages/sync-actions/src/zones-actions.js [hunk 1]
```diff
@@ -25,20 +25,45 @@ export function actionsMapLocations(diff, oldObj, newObj) {
...
-        location: oldLocation,
-      },
-      {
-        action: 'addLocation',
-        location: newLocation,
-      },
-    ],
+    [REMOVE_ACTIONS]: oldLocation =>
+      // We only add the action if the location is not included in the new object.
+      !newObj.locations.find(
```
Comment (tdeekens) [inline packages/sync-actions/src/zones-actions.js]: Using `.some` might be more explanatory.

File: packages/sync-actions/src/zones-actions.js [hunk 2]
```diff
@@ -25,20 +25,45 @@ export function actionsMapLocations(diff, oldObj, newObj) {
...
+            location: oldLocation,
+          }
+        : null,
+    [CHANGE_ACTIONS]: (oldLocation, newLocation) => {
+      const result = []
+
+      // We only remove the location in case that the oldLocation is not
+      // included in the new object
+      if (
+        !newObj.locations.find(
```
Comment (tdeekens) [inline packages/sync-actions/src/zones-actions.js]: Same.

File: packages/sync-actions/src/zones-actions.js [hunk 3]
```diff
@@ -25,20 +25,45 @@ export function actionsMapLocations(diff, oldObj, newObj) {
...
+      )
+        result.push({
+          action: 'removeLocation',
+          location: oldLocation,
+        })
+
+      // We only add the location in case that the newLocation was not
+      // included in the old object
+      if (
+        !oldObj.locations.find(
```
Comment (tdeekens) [inline packages/sync-actions/src/zones-actions.js]: Having the same thing three times I wonder if we can extract it into a

```js
const hasLocation(locations, otherLocation) =>
   locations.some(location => location.country === otherLocation.country
```

and just use it everywhere like

```js
if(!hasLocation(oldObj.locations, newLocation) {}
```

Comment (qmateub) [reply packages/sync-actions/src/zones-actions.js]: Uh! Good point 👍

File: packages/sync-actions/src/zones-actions.js (hunk 2 above)
Comment (adnasa) [reply packages/sync-actions/src/zones-actions.js]: even go as far as to put this logic into a util

File: packages/sync-actions/src/zones-actions.js (hunk 3 above)
Comment (adnasa) [reply packages/sync-actions/src/zones-actions.js]: yeah, as I mentioned in my previous comment

Comment (qmateub) [reply packages/sync-actions/src/zones-actions.js]: a180141 😎

**Pull request merged**